enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,160 @@
+    (.getPath f)))
+
+;;; Request callbacks
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [data callback]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (if *in-tests*
+        (callback)
+        (future (callback))))
+
+    ;; mt/user-http-request goes here
+    clojure.java.io.IOFactory
+    (make-input-stream [_ _]
+      (let [res (io/input-stream (if (instance? File data)
+                                   (ba-copy data)
+                                   data))]
+        (callback)
+        res))))
+
+;;; Logic
+
+(defn- serialize&pack ^File [opts]
+  (let [id       (format "%s-%s"
+                         (u/slugify (public-settings/site-name))
+                         (u.date/format "YYYY-MM-dd_HH-mm" (t/local-date-time)))
+        path     (io/file parent-dir id)
+        dst      (io/file (str (.getPath path) ".tar.gz"))
+        log-file (io/file path "export.log")]
+    (with-open [_logger (logger/for-ns 'metabase-enterprise.serialization log-file)]
+      (try                              ; try/catch inside logging to log errors
+        (serdes/with-cache
+          (-> (extract/extract opts)
+              (storage/store! path)))
+        ;; not removing storage immediately to save some time before response
+        (u.compress/tgz path dst)
+        (catch Exception e
+          (log/error e "Error during serialization"))))
+    {:archive  (when (.exists dst)
+                 dst)
+     :log-file (when (.exists log-file)
+                 log-file)
+     :callback (fn []
+                 (when (.exists path)
+                   (run! io/delete-file (reverse (file-seq path))))
+                 (when (.exists dst)
+                   (io/delete-file dst)))}))
+
+(defn- unpack&import [^File file & [size]]
+  (let [dst      (io/file parent-dir (u.random/random-name))
+        log-file (io/file dst "import.log")]
+    (with-open [_logger (logger/for-ns 'metabase-enterprise.serialization log-file)]
+      (try                              ; try/catch inside logging to log errors
+        (log/infof "Serdes import, size %s" size)
+        (let [path (u.compress/untgz file dst)]
+          (serdes/with-cache
+            (-> (v2.ingest/ingest-yaml (.getPath (io/file dst path)))
+                (v2.load/load-metabase! {:abort-on-error true}))))
+        (catch Exception e
+          (log/error e "Error during serialization"))))
+    {:log-file log-file
+     :callback #(when (.exists dst)
+                  (run! io/delete-file (reverse (file-seq dst))))}))
+
+(comment
+  ;; dump on disk and see that it works
+  (serialize&pack {:no-data-model true :no-settings true :targets [["Collection" 31409]]})
+  (unpack&import (io/file "metabase_test.tar.gz")))
```

Comment:
do you mean to keep this? it's kind of random IMO

Reply:
Eh, it kind of is, but I used it so many times that it felt useful to keep it in. :) But yeah, after a pause looks too random. 😁